File: Languages/python/algorithms-dsa/leetcode/linked-list/reverse_linked_list.py
"""
LeetCode #206 - Reverse Linked List
Topic: Linked List
Difficulty: Easy

Reverse a singly linked list iteratively.

Time Complexity: O(n)
Space Complexity: O(1)
"""


class ListNode:
    def __init__(self, val: int = 0, next: "ListNode | None" = None):
        self.val = val
        self.next = next


def reverse_list(head: ListNode | None) -> ListNode | None:
    prev = None
    curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev


def to_list(head: ListNode | None) -> list[int]:
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result


def from_list(vals: list[int]) -> ListNode | None:
    dummy = ListNode(0)
    curr = dummy
    for v in vals:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


if __name__ == "__main__":
    assert to_list(reverse_list(from_list([1, 2, 3, 4, 5]))) == [5, 4, 3, 2, 1]
    assert to_list(reverse_list(from_list([1, 2]))) == [2, 1]
    assert to_list(reverse_list(from_list([]))) == []
    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/leetcode/linked-list/reverse_linked_list.cpp
/*
 * LeetCode 206 - Reverse Linked List
 * Topic: Linked List
 * Difficulty: Easy
 *
 * Iterative pointer swap: prev, curr, next.
 * Time: O(n)
 * Space: O(1)
 */
#include <iostream>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* reverseList(ListNode* head) {
    ListNode* prev = nullptr;
    while (head) {
        ListNode* next = head->next;
        head->next = prev;
        prev = head;
        head = next;
    }
    return prev;
}

ListNode* buildList(initializer_list<int> vals) {
    ListNode dummy(0);
    ListNode* curr = &dummy;
    for (int v : vals) { curr->next = new ListNode(v); curr = curr->next; }
    return dummy.next;
}

bool equals(ListNode* a, initializer_list<int> vals) {
    for (int v : vals) { if (!a || a->val != v) return false; a = a->next; }
    return a == nullptr;
}

int main() {
    assert(equals(reverseList(buildList({1,2,3,4,5})), {5,4,3,2,1}));
    assert(equals(reverseList(buildList({1,2})), {2,1}));
    assert(reverseList(nullptr) == nullptr);

    cout << "All tests passed!" << endl;
    return 0;
}
